docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1060 @@
+
+While `Vector256.Shuffle` treats it as a "single 256-bit vector" (rather than "2x128-bit vectors"). This was done for consistency and to better map to a cross-platform mentality where `AVX-512` and `SVE` all operate on "full width".
+
+## Summary
+
+The main goal of the new `Vector128` and `Vector256` APIs is to make writing fast, vectorized code possible without becoming familiar with hardware-specific instructions and becoming an assembly language expert. Our recommendations depend on your current expertise level, software you maintain and the one you need to create:
+
+- If you are already an expert and you have vectorized your code for both `x64/x86` and `arm64/arm` code you can use the new APIs to simplify your code, but you most likely won't observe any performance gains. [#64451](https://github.com/dotnet/runtime/issues/64451) lists the places where it was/can be done in dotnet/runtime. You can use links to the merged PRs to see real-life examples.
+- If you have already vectorized your code, but only for `x64/x86` or `arm64/arm`, you can use the  new APIs to have a single, cross-platform implementation.
+- If you have already vectorized your code with `Vector<T>` you can use the new APIs to check if they can produce better codegen.
+- If you are not familiar with hardware specific instructions or you are about to vectorize a scalar algorithm, you should start with the new `Vector128` and `Vector256` APIs. Get a solid and working implementation and eventually consider using hardware-specific methods for performance critical code paths.
+
+### Best practices
+
+1. Implement tests that cover all code paths, including Acces Violation.
+2. Run tests for all hardware acceleration scenarios, use the existing env vars to do that.
+3. Implement benchmarks that mimic real life scenarios, do not increase the complexity of your code when it's not beneficial for your end users.
+4. Prefer managed references over unsafe pointers to avoid pinning and safety issues.
+5. Use `ref MemoryMarshal.GetReference(span)` instead `ref span[0]` and `ref MemoryMarshal.GetArrayDataReference(array)` instead `ref array[0]` to handle empty buffers correctly.
+6. Prefer `LoadUnsafe(ref T, nuint elementOffset)` and `StoreUnsafe(this Vector128<T> source, ref T destination, nuint elementOffset)` over other methods for loading and storing vectors as they avoid pinning and the need of doing pointer arithmetic. Be aware of unsigned integer overflow!
+7. Always handle the vectorized loop remainder.
+8. When storing values in memory, be aware of a potential buffer overlap.
+9. When writing a vectorized algorithm, start with writing the tests for edge cases, then implement a scalar solution and afterwards try to express what the scalar code is doing with Vector128/256 APIs. Over time, you may gain enough experience to skip the scalar step.
```

Comment:
```suggestion
9. When writing a vectorized algorithm, start with writing the tests for edge cases, then implement a scalar solution and afterwards try to express what the scalar code is doing with Vector128/256 APIs.
```
As this document is focused on beginners, they shouldn't be encouraged to skip the scalar step.
They can figure out that way by themself once they got experienced enough 😉.

Reply:
Let's wait for other reviewers to chime in, as I am not convinced yet.

Reply:
I would say you should never skip the scalar step.

You should always start with a solid scalar baseline. This scalar implementation is ultimately required since not all platforms will support vectorization (and developers can turn off acceleration with an environment variable). It also provides a nice foundation to write your tests against, to benchmark against, and to validate correctness against.

The next logical step is to unroll the operation to do `Vector###<T>.Count` operations at a time. This step, even without vectorization can provide its own perf benefits as it can often take advantage of pipelining in the CPU. -- This is the only step, if any, that I would consider "skippable" and that just comes with practice/experience.

Then you squash those unrolled operations into a single `Vector###<T>` operation to end up with a vectorized implementation.

You can finally and optionally, do additional tweaks to improve the vectorized implementation such as handling alignment, more efficiently processing trailing elements, or doing additional unrolling of the vectorized code to take advantage of SIMD pipelining.